compiler/optimizer/LocalOpts.cpp

```diff
@@ -6932,183 +6928,108 @@ int32_t TR_InvariantArgumentPreexistence::perform()
          {
-         if (trace())
-            traceMsg(comp(), "PREX:    Populating parmInfo of inlined method %s from argInfo %p\n", feMethod->signature(trMemory()), argInfo);
+         traceIfEnabled("PREX:    Populating parmInfo of inlined method %s from argInfo %p\n", feMethod->signature(trMemory()), argInfo);
 
          TR_PrexArgInfo *argInfo = comp()->getCurrentInlinedCallArgInfo();
-         int32_t index = 0;
+         index = 0;
          for (TR::ParameterSymbol *p = parms.getFirst(); p != NULL; p = parms.getNext(), index++)
             {
             TR_PrexArgument *arg = argInfo->get(index);
             ParmInfo &parmInfo = _parmInfo[index];
-            if (arg)
+            TR_OpaqueClassBlock *clazzFromMethod = parmInfo.getClass();
+
+            if (!parmInfo.isInvariant())
                {
-               if (trace())
-                  traceMsg(comp(), "PREX:      Parm %d is arg %p parmInfo %p\n", index, arg, &parmInfo);
+               // Clear non-invariant arg info
+               argInfo->set(index, NULL);
+               traceIfEnabled("PREX:       parm %d is not invariant\n", index);
+               continue;
                }
-            else
+
+            if (parmInfo.hasKnownObjectIndex() || !clazzFromMethod)
+               continue;
+
+            if (!arg)
                {
-               if (trace())
-                  traceMsg(comp(), "PREX:      No argInfo for parm %d\n", index);
+               traceIfEnabled("PREX:      No argInfo for parm %d\n", index);
                continue;
                }
 
+            traceIfEnabled("PREX:      Parm %d is arg %p parmInfo %p\n", index, arg, &parmInfo);
+
+            if (arg->hasKnownObjectIndex() && !(arg->getClass() && arg->classIsFixed()))
+               {
+               TR::VMAccessCriticalSection setClass(comp());
+               TR_OpaqueClassBlock *fixedClazz = TR::Compiler->cls.objectClass(comp(), knot->getPointer(arg->getKnownObjectIndex()));
+               arg->setClassIsFixed(fixedClazz);
+               traceIfEnabled("PREX:        Parm %d is known object obj%d\n", index, arg->getKnownObjectIndex());
+               }
+
             bool classIsFixed       = arg->classIsFixed();
             bool classIsPreexistent = arg->classIsPreexistent();
+            if (!classIsFixed && !classIsPreexistent)
+               continue;
 
-            if (classIsFixed || classIsPreexistent)
+            TR_OpaqueClassBlock *clazz = arg->getClass() ? arg->getClass() : clazzFromMethod;
+
+            // bail out if types are not compatible, which is an indicator that we're peeking or inlining dead path
+            if (clazz != clazzFromMethod && fe()->isInstanceOf(clazz, clazzFromMethod, true, true, true) != TR_yes)
+               return 1;
+
+            parmInfo.clear();
+            parmInfo.setSymbol(p);
+            parmInfo.setKnownObjectIndex(arg->getKnownObjectIndex());
+            parmInfo.setClass(clazz);
+            char *sig = NULL;
```

Comment:
I think this is pretty ugly since it is tracing code littering the mainline - It feels to me like we should only do it on the tracing paths just for code maintenance - the small extra memory consumption is not worth the mainline complexity IMO. If you feel strongly otherwise please provide a justification so we can discuss the pros and cons of our different view points.

Reply:
Moved to tracing path.